ouroboros-consensus/test-storage/Test/Ouroboros/Storage/ChainDB/VolDB.hs

```diff
@@ -0,0 +1,340 @@
+
+-- | Newtype that uses 'ppShow' for its 'Show' instance.
+newtype Pretty a = Pretty a
+  deriving (Eq, Arbitrary)
+
+instance Show a => Show (Pretty a) where
+  show (Pretty a) = ppShow a
+
+{-------------------------------------------------------------------------------
+  Generators
+-------------------------------------------------------------------------------}
+
+-- | All blocks on the current chain use this value as for their body. This
+-- allows us to statically convert a header from the current chain to a block.
+currentChainBody :: TestBody
+currentChainBody = TestBody 0 True
+
+genFirstBlock ::
+     Gen TestBody
+  -> Gen TestBlock
+genFirstBlock genBody = oneof
+    [ firstBlock 0            <$> genBody
+    , firstEBB   (const True) <$> genBody
+    ]
+
+genSuccessor ::
+     TestBlock
+  -> Gen TestBody
+  -> Gen TestBlock
+genSuccessor prevBlk genBody = frequency
+    [ ( if prevIsEBB then 0 else 1
+      , mkNextEBB (const True) prevBlk
+          <$> chooseSlot (prevSlot + 1) (prevSlot + 2)
+          -- We don't care about EpochNo
+          <*> return 0
+          <*> genBody
+      )
+    , (3
+      , mkNextBlock prevBlk
+          <$> chooseSlot (prevSlot + if prevIsEBB then 0 else 1) (prevSlot + 2)
+          <*> genBody
+      )
+    ]
+  where
+    prevSlot  = blockSlot prevBlk
+    prevIsEBB = isJust $ headerIsEBB $ getHeader prevBlk
+
+    chooseSlot :: SlotNo -> SlotNo -> Gen SlotNo
+    chooseSlot (SlotNo start) (SlotNo end) = SlotNo <$> choose (start, end)
+
+genChainHelper ::
+     Gen TestBody
+  -> Int                                  -- ^ Number of headers to generate
+  -> TestBlock                            -- ^ Previous block
+  -> AnchoredFragment (Header TestBlock)  -- ^ Accumulator
+  -> Gen (AnchoredFragment (Header TestBlock))
+genChainHelper genBody n prevBlk frag = case n of
+    0 -> return frag
+    _ -> do
+      nextBlk <- genSuccessor prevBlk genBody
+      genChainHelper genBody (n - 1) nextBlk (frag AF.:> getHeader nextBlk)
+
+-- | Also returns the header corresponding to the anchor, unless the anchor is
+-- genesis.
+generateChain :: Gen ( AnchoredFragment (Header TestBlock)
+                     , Maybe (Header TestBlock)
+                     )
+generateChain = sized $ \size -> do
+    -- + 1 because the chain is size + 1 long (we generate the first block
+    -- separately)
+    takeNewestN <- choose (0, size + 1)
+    firstBlk    <- genFirstBlock (return currentChainBody)
+    let initFragment = AF.Empty AF.AnchorGenesis AF.:> getHeader firstBlk
+    fullChain <- genChainHelper
+                   (return currentChainBody)
+                   size
+                   firstBlk
+                   initFragment
+    let mbAnchorHdr = case AF.head (AF.dropNewest takeNewestN fullChain) of
+                        Left _          -> Nothing
+                        Right anchorHdr -> Just anchorHdr
+    return (AF.anchorNewest (fromIntegral takeNewestN) fullChain, mbAnchorHdr)
```

Comment:
TODO: never an empty chain anchored at genesis, i.e., drop the `firstBlk`. (case at the start)